### PR TITLE
EAHW-2545: Rework transactions and handlermappingconfig

### PR DESCRIPTION
**Pull request by daleelR**
This PR is focused around improving the transaction handling and simplifying the bulky HandlerMappingConfig class.

### Transactions
- Using the Spring TransactionTemplate lambda to wrap the relevant service logic in transactions
- Attempted to use @Transactional annotation but this didn't work as our service classes are 'self managed' and injected into the app context and runtime
  - It will only work for beans that are purely handled by spring with no manual dependency injection

### BDD
- Decode filepath to handle special characters used for file path spaces

### HandlerMappingConfig Class
- Shifted the baseEntity checking logic into separate service
- Shifted the controller endpoint registering code into separate service
- Tidied up the Resource OpenApi path discovery code

### Other
- Register controller as a spring bean
- Service factory now creates the repo for the service instead of the HanderMappingConfig
  - This also allows removing the EntityManager dependency from HanderMappingConfig making it unit testable

### Tests
- Made HandlerMappingConfigTest unit testable with more coverage
  - Shifted existing tests into new ControllerRegistererServiceTest
- Service & Controller Factory tests are now integration tests to ensure beans are actually registered in the context
- Exclude fields with circular relationships from equals and hashcode on demo/test models to fix querying loop issue

### CODE REVIEW

**Review by JWeston-HO — DISMISSED** [review]

**Review by taomoh — COMMENTED** [review]
Great effort @daleelR . I have a few minor comments

File: jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/ResourceApiService.java [hunk 1]
```diff
@@ -29,18 +29,18 @@
...
-                            PlatformTransactionManager transactionManager,
                             TenantRepository<T> repository,
-                            EntityValidator entityValidator) {
+                            EntityValidator entityValidator,
+                            PlatformTransactionManager platformTransactionManager) {
     this.entityUtils = entityUtils;
-    this.transactionManager = transactionManager;
     this.repository = repository;
     this.entityValidator = entityValidator;
+    this.transactionTemplate = new TransactionTemplate(platformTransactionManager);
```
[the remark this answers is not in the document]
Comment (daleelR) [reply jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/ResourceApiService.java]: As discussed have now instantiated the transactionTemplate once in the factory and passed it in to all it's created services.

File: jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/BaseEntityCheckerService.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  private boolean classHasBaseEntityParent(Class<?> childClass) {
+    var superType = childClass.getSuperclass();
+    while (!superType.equals(Object.class)) {
+      if (superType.equals(BaseEntity.class)) {
+        return true;
+      }
+      superType = superType.getSuperclass();
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (daleelR) [reply jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/BaseEntityCheckerService.java]: Ah I didn't realise there was a built in method for this. Have updated.

File: jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/BaseEntityCheckerService.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  /**
+   * Filters all the entities and returns a map of only the BaseEntity subclasses
+   *
+   * @return A map of the key as the entity java class and the value as the entity name
+   */
+  public Map<Class<?>, String> filterBaseEntitySubClasses() {
+    return entityManager.getMetamodel().getEntities().stream()
+        .filter(entityType -> entityType.getJavaType().isAnnotationPresent(Resource.class)
+            && classHasBaseEntityParent(entityType.getJavaType()))
```
[the remark this answers is not in the document]
Comment (daleelR) [reply jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/service/BaseEntityCheckerService.java]: Yep agreed, that does look better. Have updated.

File: jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/factory/ResourceApiControllerFactory.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  public ResourceApiControllerFactory(ObjectMapper objectMapper,
+                                      GenericApplicationContext context) {
+    this.objectMapper = objectMapper;
+    this.context = context;
+  }
+
+
+  public <T extends BaseEntity> ResourceApiController<T> getBean(
+      Class<T> resourceClass,
+      ResourceApiService<T> resourceApiService) {
```
[the remark this answers is not in the document]
Comment (daleelR) [reply jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/factory/ResourceApiControllerFactory.java]: After the additional refactoring carried out in our pairing session, I've now merged those changes in which resolves this.

[the remark this answers is not in the document]
Comment (taomoh) [reply jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/factory/ResourceApiControllerFactory.java]: Thanks @daleelR for bearing with me :)

**Review by taomoh — APPROVED** [review]
Great effort @daleelR !

**Review by JWeston-HO — APPROVED** [review]

**Pull request merged**